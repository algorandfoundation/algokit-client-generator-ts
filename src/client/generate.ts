import { DocumentParts, inline, NewLine } from '../output/writer'
import { paramsFactory } from './params-factory'
import { appClient } from './app-client'
import { deployTypes } from './deploy-types'
import { utilityTypes } from './utility-types'
import { imports } from './imports'
import { createGeneratorContext, GeneratorOptions } from './generator-context'
import { appTypes } from './app-types'
import { callComposerType } from './call-composer-types'
import { Arc56Contract, StructField } from '@algorandfoundation/algokit-utils/types/app-arc56'
import { appFactory } from './app-factory'
import { Sanitizer } from '../util/sanitization'

function convertStructs(s: StructField[], sanitizer: Sanitizer): StructField[] {
  return s.map(
    ({ name, type }) =>
      ({
        name: sanitizer.makeSafePropertyIdentifier(name),
        type: typeof type === 'string' ? type : convertStructs(type, sanitizer),
      }) satisfies StructField,
  )
}

// TODO: NC - This can be cleaned up further
function shrinkAppSpec(app: Arc56Contract, isSlim: boolean = false): Arc56Contract {
  const strippedApp = { ...app } satisfies Arc56Contract

  if (strippedApp.sourceInfo?.approval?.sourceInfo) {
    // Keep only source info entries that have errorMessage
    strippedApp.sourceInfo.approval.sourceInfo = strippedApp.sourceInfo.approval.sourceInfo
      .filter((entry) => entry.errorMessage)
      .map((entry) => ({
        pc: entry.pc,
        errorMessage: entry.errorMessage,
        // Keep minimal context for error mapping if available
        ...(entry.teal !== undefined && { teal: entry.teal }),
      }))
  }

  if (strippedApp.sourceInfo?.clear?.sourceInfo) {
    // Apply same logic to clear program source info
    strippedApp.sourceInfo.clear.sourceInfo = strippedApp.sourceInfo.clear.sourceInfo
      .filter((entry) => entry.errorMessage)
      .map((entry) => ({
        pc: entry.pc,
        errorMessage: entry.errorMessage,
        // Keep minimal context for error mapping if available
        ...(entry.teal !== undefined && { teal: entry.teal }),
      }))
  }

  if (strippedApp.compilerInfo) {
    delete strippedApp.compilerInfo
  }

  // These are needed for deployment but not for calling existing contracts
  if (isSlim) {
    if (strippedApp.source) {
      delete strippedApp.source
    }
    if (strippedApp.byteCode) {
      delete strippedApp.byteCode
    }
    if (strippedApp.templateVariables) {
      delete strippedApp.templateVariables
    }
    if (strippedApp.scratchVariables) {
      delete strippedApp.scratchVariables
    }
  }
  return strippedApp
}

export function* generate(app: Arc56Contract, options: GeneratorOptions = { preserveNames: false }): DocumentParts {
  // Apply source info stripping if slim mode is enabled
  const reduceAppSpec = shrinkAppSpec(app, options.slim)

  const ctx = createGeneratorContext(reduceAppSpec, options)
  yield `/* eslint-disable */`
  yield `/**`
  yield ` * This file was automatically generated by @algorandfoundation/algokit-client-generator.`
  yield ` * DO NOT MODIFY IT BY HAND.`
  yield ` * requires: @algorandfoundation/algokit-utils: ^7`
  yield ` */`

  yield* imports(ctx)
  // Change the structs definition to sanitize property names according to the defined rules
  // for instance, this may (unless you passed in --preserve-names) convert properties like my_prop to myProp
  reduceAppSpec.structs = Object.fromEntries(
    Object.keys(reduceAppSpec.structs).map((key) => [key, convertStructs(reduceAppSpec.structs[key], ctx.sanitizer)]),
  )
  yield* inline('export const APP_SPEC: Arc56Contract = ', JSON.stringify(reduceAppSpec), ' as unknown as Arc56Contract')
  yield NewLine

  yield* utilityTypes()
  yield NewLine
  yield* appTypes(ctx)

  // Only generate deploy types if not in slim mode
  if (!ctx.slim) {
    yield* deployTypes(ctx)
  }
  yield NewLine

  // Write a call factory
  yield* paramsFactory(ctx)
  yield NewLine
  // Write a factory only if not in slim mode
  if (!ctx.slim) {
    yield* appFactory(ctx)
  }
  // Write a client
  yield* appClient(ctx)

  yield* callComposerType(ctx)
}
