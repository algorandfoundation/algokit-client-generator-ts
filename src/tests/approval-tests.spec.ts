import fs from 'fs'
import path from 'path'
import { generate } from '../client/generate.js'
import { writeDocumentPartsToString } from '../output/writer.js'
import { loadApplicationJson } from '../schema/load.js'
import { expect, test, describe } from 'vitest'
import { pascalCase } from "change-case";

const writeActual = process.env.TEST_ENV !== 'ci'

const arc32TestContracts = [
  'duplicate_structs', 'hello_world', 'life_cycle', 'minimal', 'state', 'voting_round'
] as const
const arc56TestContracts = [
  'reti', 'arc56_test', 'nested', 'nfd', 'structs', 'zero_coupon_bond'
] as const
const testContracts = [
  ...arc32TestContracts.map((app) => [app, 'arc32']),
  ...arc56TestContracts.map((app) => [app, 'arc56']),
]

describe('When generating a ts client for a the contract', () => {
  test.each(testContracts)('%s approval', async (contractName, extension) => {
    const dir = path.join(__dirname, `../../examples/smart_contracts/artifacts/${contractName}/`)
    const spec = await loadApplicationJson(path.join(dir, `${pascalCase(contractName)}.${extension}.json`))

    const result = writeDocumentPartsToString(generate(spec))
    if (writeActual) fs.writeFileSync(path.join(dir, `client.generated.ts`), result)

    const approvedClient = fs.readFileSync(path.join(dir, `client.ts`), 'utf-8')
    expect(result).toBe(approvedClient)
  })
})
