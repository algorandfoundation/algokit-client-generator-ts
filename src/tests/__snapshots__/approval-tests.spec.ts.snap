// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Client type differences > arc56_test: full vs minimal differences > arc56_test-client-diff 1`] = `
[
  "appClient._appSpec.scratchVariables",
  "appClient._appSpec.scratchVariables.someNumber",
  "appClient._appSpec.scratchVariables.someNumber.slot",
  "appClient._appSpec.scratchVariables.someNumber.type",
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "appClient._appSpec.templateVariables",
  "appClient._appSpec.templateVariables.someNumber",
  "appClient._appSpec.templateVariables.someNumber.type",
]
`;

exports[`Client type differences > arc56_test: full vs minimal differences > arc56_test-composer-diff 1`] = `[]`;

exports[`Client type differences > duplicate_structs: full vs minimal differences > duplicate_structs-client-diff 1`] = `
[
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
]
`;

exports[`Client type differences > duplicate_structs: full vs minimal differences > duplicate_structs-composer-diff 1`] = `[]`;

exports[`Client type differences > hello_world: full vs minimal differences > hello_world-client-diff 1`] = `
[
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "createTransaction.delete",
  "createTransaction.delete.bare",
  "createTransaction.update",
  "createTransaction.update.bare",
  "params.delete",
  "params.delete.bare",
  "params.update",
  "params.update.bare",
  "send.delete",
  "send.delete.bare",
  "send.update",
  "send.update.bare",
]
`;

exports[`Client type differences > hello_world: full vs minimal differences > hello_world-composer-diff 1`] = `
[
  "delete",
  "delete.bare",
  "update",
  "update.bare",
]
`;

exports[`Client type differences > life_cycle: full vs minimal differences > life_cycle-client-diff 1`] = `
[
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "createTransaction.delete",
  "createTransaction.delete.deleteTest",
  "createTransaction.update",
  "createTransaction.update.bare",
  "createTransaction.update.updateTest",
  "params.delete",
  "params.delete.deleteTest",
  "params.update",
  "params.update.bare",
  "params.update.updateTest",
  "send.delete",
  "send.delete.deleteTest",
  "send.update",
  "send.update.bare",
  "send.update.updateTest",
]
`;

exports[`Client type differences > life_cycle: full vs minimal differences > life_cycle-composer-diff 1`] = `
[
  "delete",
  "delete.deleteTest",
  "update",
  "update.bare",
  "update.updateTest",
]
`;

exports[`Client type differences > minimal: full vs minimal differences > minimal-client-diff 1`] = `
[
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "createTransaction.delete",
  "createTransaction.delete.bare",
  "createTransaction.update",
  "createTransaction.update.bare",
  "params.delete",
  "params.delete.bare",
  "params.update",
  "params.update.bare",
  "send.delete",
  "send.delete.bare",
  "send.update",
  "send.update.bare",
]
`;

exports[`Client type differences > minimal: full vs minimal differences > minimal-composer-diff 1`] = `
[
  "delete",
  "delete.bare",
  "update",
  "update.bare",
]
`;

exports[`Client type differences > nested: full vs minimal differences > nested-client-diff 1`] = `
[
  "appClient._appSpec.byteCode",
  "appClient._appSpec.byteCode.approval",
  "appClient._appSpec.byteCode.clear",
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "appClient._appSpec.templateVariables",
]
`;

exports[`Client type differences > nested: full vs minimal differences > nested-composer-diff 1`] = `[]`;

exports[`Client type differences > nfd: full vs minimal differences > nfd-client-diff 1`] = `
[
  "appClient._appSpec.scratchVariables",
  "appClient._appSpec.scratchVariables.adminAsaId",
  "appClient._appSpec.scratchVariables.adminAsaId.slot",
  "appClient._appSpec.scratchVariables.adminAsaId.type",
  "appClient._appSpec.scratchVariables.registryAppId",
  "appClient._appSpec.scratchVariables.registryAppId.slot",
  "appClient._appSpec.scratchVariables.registryAppId.type",
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "appClient._appSpec.templateVariables",
  "appClient._appSpec.templateVariables.adminAsaId",
  "appClient._appSpec.templateVariables.adminAsaId.type",
  "appClient._appSpec.templateVariables.registryAppId",
  "appClient._appSpec.templateVariables.registryAppId.type",
  "createTransaction.update",
  "createTransaction.update.updateApplication",
  "params.update",
  "params.update.updateApplication",
  "send.update",
  "send.update.updateApplication",
]
`;

exports[`Client type differences > nfd: full vs minimal differences > nfd-composer-diff 1`] = `
[
  "update",
  "update.updateApplication",
]
`;

exports[`Client type differences > reti: full vs minimal differences > reti-client-diff 1`] = `
[
  "appClient._appSpec.scratchVariables",
  "appClient._appSpec.scratchVariables.nfdRegistryAppId",
  "appClient._appSpec.scratchVariables.nfdRegistryAppId.slot",
  "appClient._appSpec.scratchVariables.nfdRegistryAppId.type",
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "appClient._appSpec.templateVariables",
  "appClient._appSpec.templateVariables.nfdRegistryAppId",
  "appClient._appSpec.templateVariables.nfdRegistryAppId.type",
]
`;

exports[`Client type differences > reti: full vs minimal differences > reti-composer-diff 1`] = `[]`;

exports[`Client type differences > state: full vs minimal differences > state-client-diff 1`] = `
[
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "createTransaction.delete",
  "createTransaction.delete.bare",
  "createTransaction.delete.deleteAbi",
  "createTransaction.update",
  "createTransaction.update.bare",
  "createTransaction.update.updateAbi",
  "params.delete",
  "params.delete.bare",
  "params.delete.deleteAbi",
  "params.update",
  "params.update.bare",
  "params.update.updateAbi",
  "send.delete",
  "send.delete.bare",
  "send.delete.deleteAbi",
  "send.update",
  "send.update.bare",
  "send.update.updateAbi",
]
`;

exports[`Client type differences > state: full vs minimal differences > state-composer-diff 1`] = `
[
  "delete",
  "delete.bare",
  "delete.deleteAbi",
  "update",
  "update.bare",
  "update.updateAbi",
]
`;

exports[`Client type differences > structs: full vs minimal differences > structs-client-diff 1`] = `
[
  "appClient._appSpec.byteCode",
  "appClient._appSpec.byteCode.approval",
  "appClient._appSpec.byteCode.clear",
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "appClient._appSpec.templateVariables",
]
`;

exports[`Client type differences > structs: full vs minimal differences > structs-composer-diff 1`] = `[]`;

exports[`Client type differences > void: full vs minimal differences > void-client-diff 1`] = `
[
  "appClient._appSpec.byteCode",
  "appClient._appSpec.byteCode.approval",
  "appClient._appSpec.byteCode.clear",
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "appClient._appSpec.templateVariables",
  "appClient._appSpec.templateVariables.DELETABLE",
  "appClient._appSpec.templateVariables.DELETABLE.type",
  "appClient._appSpec.templateVariables.UPDATABLE",
  "appClient._appSpec.templateVariables.UPDATABLE.type",
  "createTransaction.delete",
  "createTransaction.delete.bare",
  "createTransaction.update",
  "createTransaction.update.bare",
  "params.delete",
  "params.delete.bare",
  "params.update",
  "params.update.bare",
  "send.delete",
  "send.delete.bare",
  "send.update",
  "send.update.bare",
]
`;

exports[`Client type differences > void: full vs minimal differences > void-composer-diff 1`] = `
[
  "delete",
  "delete.bare",
  "update",
  "update.bare",
]
`;

exports[`Client type differences > voting_round: full vs minimal differences > voting_round-client-diff 1`] = `
[
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "createTransaction.delete",
  "createTransaction.delete.bare",
  "params.delete",
  "params.delete.bare",
  "send.delete",
  "send.delete.bare",
]
`;

exports[`Client type differences > voting_round: full vs minimal differences > voting_round-composer-diff 1`] = `
[
  "delete",
  "delete.bare",
]
`;

exports[`Client type differences > zero_coupon_bond: full vs minimal differences > zero_coupon_bond-client-diff 1`] = `
[
  "appClient._appSpec.byteCode",
  "appClient._appSpec.byteCode.approval",
  "appClient._appSpec.byteCode.clear",
  "appClient._appSpec.source",
  "appClient._appSpec.source.approval",
  "appClient._appSpec.source.clear",
  "appClient._appSpec.templateVariables",
  "createTransaction.update",
  "createTransaction.update.bare",
  "params.update",
  "params.update.bare",
  "send.update",
  "send.update.bare",
]
`;

exports[`Client type differences > zero_coupon_bond: full vs minimal differences > zero_coupon_bond-composer-diff 1`] = `
[
  "update",
  "update.bare",
]
`;
